### PR TITLE
Extend rankUpdate! method to allow for upper tri

### PR DESCRIPTION
**Pull request by dmbates**
- Add checks of `require_one_based_indexing` and macro calls `@inbounds` in some files in the `linalg` directory

- Add more checks for positive definiteness in `cholUnblocked.jl`.

- Although this commit allows for `rankUpdate!` of the upper triangle it turns out that this is much slower than updating the lower triangle.

```julia
julia> using BenchmarkTools, LinearAlgebra, MixedModels
[ Info: Precompiling MixedModels [ff71e718-51f3-5ec2-a782-8ffcbfa3c316]

julia> const contrasts = Dict{Symbol,Any}(:subj => Grouping(), :s => Grouping(), :d => Grouping(), :dept => Grouping(), :service => EffectsCoding());

julia> const form = @formula(y ~ 1 + service + (1|s) + (1|d) + (1|dept));

julia> insteval = MixedModels.dataset(:insteval);

julia> iem01 = fit(MixedModel, form, insteval; contrasts)
Minimizing 115 	 Time: 0:00:01 (16.15 ms/it)
Linear mixed model fit by maximum likelihood
 y ~ 1 + service + (1 | s) + (1 | d) + (1 | dept)
    logLik     -2 logLik       AIC         AICc          BIC     
 -118860.8844  237721.7688  237733.7688  237733.7699  237788.9926

Variance components:
            Column    Variance  Std.Dev. 
s        (Intercept)  0.1059724 0.3255340
d        (Intercept)  0.2652055 0.5149811
dept     (Intercept)  0.0061676 0.0785341
Residual              1.3864886 1.1774925
 Number of obs: 73421; levels of grouping factors: 2972, 1128, 14

  Fixed-effects parameters:
─────────────────────────────────────────────────────
                  Coef.  Std. Error       z  Pr(>|z|)
─────────────────────────────────────────────────────
(Intercept)   3.23629    0.0281513   114.96    <1e-99
service: Y   -0.0462943  0.00669159   -6.92    <1e-11
─────────────────────────────────────────────────────

julia> @time fit(MixedModel, form, insteval; contrasts);
Minimizing 115 	 Time: 0:00:00 ( 8.68 ms/it)
  1.082987 seconds (310.67 k allocations: 52.726 MiB, 2.81% gc time)

julia> const C = zeros(1128, 1128);

julia> const A = iem01.L[2];

julia> @benchmark MixedModels.rankUpdate!(CC, AA, 1.0, 1.0) setup=(CC=Symmetric(fill!(C, 0.0), :L); AA=A)
BenchmarkTools.Trial: 2676 samples with 1 evaluation.
 Range (min … max):  1.583 ms …  2.572 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.621 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.629 ms ± 42.664 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

        ▃▃▄▄▇▇█▆▄▄▂                                           
  ▁▁▂▃▄████████████▇▇▅▅▅▅▄▄▃▃▃▃▃▂▂▂▂▂▂▂▂▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  1.58 ms        Histogram: frequency by time        1.75 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark MixedModels.rankUpdate!(CC, AA, 1.0, 1.0) setup=(CC=Symmetric(fill!(C, 0.0), :U); AA=sparse(transpose(A)))
BenchmarkTools.Trial: 26 samples with 1 evaluation.
 Range (min … max):  193.801 ms … 194.178 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     194.037 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   194.027 ms ± 100.684 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁   ▁▁                   ▁ ███ ▁  ▁▁    ▁▁█ ▁ ▁  ▁█    ▁  █ ▁  
  █▁▁▁██▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁███▁█▁▁██▁▁▁▁███▁█▁█▁▁██▁▁▁▁█▁▁█▁█ ▁
  194 ms           Histogram: frequency by time          194 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

### CODE REVIEW

**Comment by palday** [comment]
xref: #579

**Comment by dmbates** [comment]
@palday Should this PR be merged with #637?  If so, could you do rebase one or the other - I still bear the scars of trying to rebase the julia repository when I was just learning git.

**Comment by palday** [comment]
#637 pulled in

File: src/linalg/cholUnblocked.jl [hunk 1]
```diff
@@ -9,24 +9,29 @@ because these are part of the inner calculations in a blocked Cholesky factoriza
...
 
 function cholUnblocked!(A::StridedMatrix{T}, ::Type{Val{:L}}) where {T<:BlasFloat}
     n = LinearAlgebra.checksquare(A)
     if n == 1
-        A[1] < zero(T) && throw(PosDefException(1))
+        A[1] < zero(T) && throw(LinearAlgebra.PosDefException(1))
         A[1] = sqrt(A[1])
     elseif n == 2
-        A[1] < zero(T) && throw(PosDefException(1))
+        A[1] < zero(T) && throw(LinearAlgebra.PosDefException(1))
```
Comment (palday) [inline src/linalg/cholUnblocked.jl:27]: why do we need this module qualification now?

**Review by palday — APPROVED** [review]

File: src/linalg/cholUnblocked.jl (hunk 1 above)
Comment (dmbates) [reply src/linalg/cholUnblocked.jl:27]: I had forgotten that `PosDefException` was listed in a `using` statement.  I will revert to the unqualified name in all the uses in this file.

**Pull request merged**